### PR TITLE
test: Fuzzing go 1.18

### PR DESCRIPTION
**Pull request by naveensrinivasan**
This is going to be multiple PR's for fuzzing. After this PR is merged, the goal is to get it integrated with https://github.com/google/oss-fuzz/

Release Notes:  Included go 1.18 fuzzing. 

**Types of changes**:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

**Description of the changes being introduced by the pull request**:
Included fuzz tests. 
**Please verify and check that the pull request fulfills the following requirements**:

- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by trishankatdatadog** [comment]
This is great! Can you rebase this on top of #357 soon? Thx!

**Comment by naveensrinivasan** [comment]
> This is great! Can you rebase this on top of #357 soon? Thx!

I would wait for that to be merged, and I can rebase of master.

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@
 cmd/tuf/tuf
 cmd/tuf-client/tuf-client
 .vscode
+.idea
```
Comment (znewman01) [inline .gitignore:5]: Please remove; editor-specific configuration [belongs in developers' `.gitignore` files](https://blog.martinhujer.cz/dont-put-idea-vscode-directories-to-projects-gitignore/).

(While you're at it, remove `.vscode` as well 🙂 )

**Review by znewman01 — COMMENTED** [review]
This looks great, thanks!

IIUC this is relevant to (but does not fix) https://github.com/theupdateframework/go-tuf/issues/346; just want to make sure the PR and issue link to each other.

https://github.com/theupdateframework/go-tuf/issues/346

**Comment by naveensrinivasan** [comment]
> This looks great, thanks!
> 
> IIUC this is relevant to (but does not fix) #346; just want to make sure the PR and issue link to each other.
> 
> #346

The goal is to do multiple PR’s.

**Comment by naveensrinivasan** [comment]
https://github.com/theupdateframework/go-tuf/runs/8053810526?check_suite_focus=true

This is failing because action is using version 1.17 instead of 1.19

image

**Comment by znewman01** [comment]
I think that tests will pass if you rebase after #362.

**Comment by znewman01** [comment]
I'll take a look early next week!

File: .gitignore (hunk 1 above)
Comment (naveensrinivasan) [reply .gitignore]: Took care of it. Thanks

File: pkg/keys/ecdsa_test.go [hunk 2]
```diff
@@ -41,6 +77,32 @@ func (ECDSASuite) TestUnmarshalECDSA(c *C) {
 	c.Assert(verifier.UnmarshalPublicKey(badKey), IsNil)
 }
 
+func FuzzInvalidUnmarshalECDSA(f *testing.F) {
+	f.Fuzz(func(t *testing.T, s string) {
+		if len(s) > keySize {
+			// len(s) > 40 is because 256 / 8 = 32, and 32 + 8 = 40, this is from the bitsize of the curve, we skip
+			// this because only then will we get bad data
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply pkg/keys/ecdsa_test.go:84]: The goal of that was to avoid `keySize` larger than 40 so that the fuzzer doesn't pass that data. I still don't get your comment. Could you please explain?

File: pkg/keys/ecdsa_test.go [hunk 3]
```diff
@@ -41,6 +77,32 @@ func (ECDSASuite) TestUnmarshalECDSA(c *C) {
...
+			t.Skip()
+		}
+
+		badKeyValue, err := json.Marshal(s)
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply pkg/keys/ecdsa_test.go:88]: Do you have an example of this? Also, the goal is to oss-fuzz which we can add the necessary invalid data.

File: pkg/keys/ed25519_test.go [hunk 4]
```diff
@@ -97,3 +140,17 @@ func (Ed25519Suite) TestSignVerify(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(pubKey.Verify(msg, sig), IsNil)
 }
+func FuzzSignVerfiy(f *testing.F) {
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply pkg/keys/ed25519_test.go:143]: Agreed. Thanks!

File: pkg/keys/ed25519_test.go [hunk 5]
```diff
@@ -37,6 +38,32 @@ func (Ed25519Suite) TestUnmarshalEd25519(c *C) {
...
+		if len(s) <= 32 { // ed25519 requires strings of len greater than 32
+			t.Skip()
+		}
+		c := &C{}
+
+		pub, _, err := ed25519.GenerateKey(strings.NewReader(s))
+		c.Assert(err, IsNil)
+
+		publicKey, err := json.Marshal(map[string]string{
+			"public": hex.EncodeToString(pub),
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply pkg/keys/ed25519_test.go:52]: Sorry, I don't understand your comment. Could you please explain?

File: pkg/keys/ed25519_test.go [hunk 6]
```diff
@@ -97,3 +140,17 @@ func (Ed25519Suite) TestSignVerify(c *C) {
...
+		c := &C{}
+		signer, err := GenerateEd25519Key()
+		c.Assert(err, IsNil)
+		msg := []byte(s)
+		sig, err := signer.SignMessage(msg)
+		c.Assert(err, IsNil)
+		publicData := signer.PublicData()
+		pubKey, err := GetVerifier(publicData)
+		c.Assert(err, IsNil)
+		c.Assert(pubKey.Verify(msg, sig), IsNil)
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply pkg/keys/ed25519_test.go:154]: The `sig` and `pubKey` are being generated by other `func`. How can those be fuzzed in this func? Could you please explain?

File: pkg/keys/ecdsa_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (znewman01) [reply pkg/keys/ecdsa_test.go:84]: Oops, I read the `>` in the check backwards! This is fine, feel free to resolve.

File: pkg/keys/ed25519_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (znewman01) [reply pkg/keys/ed25519_test.go:52]: Yeah, that was a little unclear, sorry. Backing up a step: what is the point of these fuzz tests?

We get nonsense input from the Go fuzzer in whatever format/type we want, and we call `UnmarshalPublicKey` with it to try to exercise weird edge cases. `UnmarshalPublicKey` can fail in a couple of ways:

1. Unmarshalling the value of `Value` into a `p256verifier`. I don't think this is ever going to fail in an interesting way, because that would be an error in `json.Decode`, which is already fuzzed extensively.
2. [`elliptic.Unmarshal(elliptic.P256(), p.PublicKey)`](https://pkg.go.dev/crypto/elliptic#Unmarshal)

So how do we feed our garbage fuzz data into `UnmarshalPublicKey` in a way that will make (2) do something interesting? I think what you're doing (making a `Value` with a legitimate JSON encoding of a `data.PublicKey`) is a great idea to exercise good code paths. But the value of `"public"` could be any `byte[]`! What should we put there? I can think of a few ideas:

1. Some bytes using `hex.EncodeToString`: this is great, because it winds up being in the expected format and possibly triggering crypto-related failures.
2. Arbitrary data: this might trigger *other* code paths

You're currently doing (1), which makes sense, as that's the format that's expected. But doing (2) as well would maybe uncover other situations we're interested in.

Hopefully that was clearer.

File: pkg/keys/ed25519_test.go [hunk 7]
```diff
@@ -37,6 +38,32 @@ func (Ed25519Suite) TestUnmarshalEd25519(c *C) {
...
+		c.Assert(err, IsNil)
+
+		badKey := &data.PublicKey{
+			Type:       data.KeyTypeEd25519,
+			Scheme:     data.KeySchemeEd25519,
+			Algorithms: data.HashAlgorithms,
+			Value:      publicKey,
+		}
+		verifier := NewEd25519Verifier()
+		c.Assert(verifier.UnmarshalPublicKey(badKey), IsNil)
```
Comment (znewman01) [inline pkg/keys/ed25519_test.go:63]: I'm starting to think that this has uncovered a bug. Shouldn't we see the "unexpected public key length" error? https://github.com/naveensrinivasan/go-tuf/blob/9f957a20ec8019e9bf16028586af7ecbdf9e51b5/pkg/keys/ed25519.go#L64

In general, I don't think every 32-byte string should be a valid ed25519 key (it may not be on the curve). Also, we don't do any crypto in `UnmarshalPublicKey`.

File: pkg/keys/ed25519_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (znewman01) [reply pkg/keys/ed25519_test.go:154]: `sig` is just a `byte[]`, so you could have the fuzzer generate it (you wouldn't be fuzzing the sign+verify flow, just verify).

`pubKey` would be a little trickier. You do fuzz generating them above, but it's unlikely that you'd get a valid `pubKey` (barring the ed25519 verifier, but that seems to be broken). So you can ignore that.

**Comment by rdimitrov** [comment]
@naveensrinivasan - hey, I've rebased your branch and it seems that the tests need to be updated.

**Comment by rdimitrov** [comment]
Closing this since it's no longer relevant to the code base, but feel free to rebase and reopen 👍
